tools/chpldef/Server.h

```diff
@@ -40,46 +39,143 @@ namespace chpldef {
+    Server* ctx_ = nullptr;
+  public:
+    ErrorHandler(Server* ctx) : ctx_(ctx) {}
+   ~ErrorHandler() = default;
+    virtual void report(chpl::Context* chapel, const Error* err) override;
+    inline const ErrorList& errors() const { return errors_; }
+    inline void clear() { errors_.clear(); }
+    inline size_t numErrors() { return errors_.size(); }
+  };
 
 private:
   State state_ = UNINIT;
   Logger logger_;
-  chpl::owned<chpl::Context> chplctx_ = nullptr;
-  int revision_;
+  chpl::owned<chpl::Context> chapel_ = nullptr;
+  Configuration config_;
+  TextRegistry textRegistry_;
+  int64_t revision_;
+  ErrorHandler* errorHandler_ = nullptr;
+
+  chpl::owned<chpl::Context> createCompilerContext();
+  bool shouldGarbageCollect();
+  bool shouldPrepareToGarbageCollect();
+
+  inline bool isLogLevel(Logger::Level level) const {
+    return logger_.level() == level;
+  }
 
 protected:
-  friend class chpldef::Shutdown;
   friend class chpldef::Initialize;
   friend class chpldef::Initialized;
+  friend class chpldef::Shutdown;
+  friend class chpldef::DidOpen;
 
   inline void setState(State state) { state_ = state; }
+  inline TextRegistry& textRegistry() { return textRegistry_; }
 
 public:
-  Server();
+  Server(Configuration config);
  ~Server() = default;
 
+  inline ErrorHandler* errorHandler() { return errorHandler_; }
   inline State state() const { return state_; }
-  inline int revision() const { return revision_; }
-  inline const chpl::Context* chplctx() const { return chplctx_.get(); }
+  inline int64_t revision() const { return revision_; }
+  inline const chpl::Context* chapel() const { return chapel_.get(); }
 
   void setLogger(Logger&& logger);
   inline Logger& logger() { return logger_; }
-  void sleep(int msec);
-
+  inline bool isLogMessage() const { return isLogLevel(Logger::MESSAGES); }
+  inline bool isLogVerbose() const { return isLogLevel(Logger::VERBOSE); }
+  inline bool isLogTrace() const { return isLogLevel(Logger::TRACE); }
   CHPLDEF_PFMT(2, 3, void message(const char* fmt, ...));
   CHPLDEF_PFMT(2, 3, void verbose(const char* fmt, ...));
   CHPLDEF_PFMT(2, 3, void trace(const char* fmt, ...));
+
+  void sleep(int msec);
+
+  enum WithChapelConfig {
+    CHPL_NO_MASK         = 0,
+    CHPL_BUMP_REVISION   = 1,
+  };
+
+  /** Execute code with controlled access to the Chapel context. */
+  template <typename F, typename ...Ns>
+  auto withChapel(WithChapelConfig c, F&& f, Ns... ns)
+  -> decltype(f(chapel_.get(), ns...)) {
+    if (shouldGarbageCollect()) chapel_->collectGarbage();
+    if (c & CHPL_BUMP_REVISION) {
+      chapel_->advanceToNextRevision(shouldPrepareToGarbageCollect());
+      ++revision_;
+    }
+    return f(chapel_.get(), ns...);
+  }
+
+  /** Execute code with controlled access to the Chapel context. */
+  template <typename F, typename ...Ns>
+  auto withChapel(F&& f, Ns... ns)
+  -> decltype(withChapel(CHPL_NO_MASK, f, ns...)) {
+    return withChapel(CHPL_NO_MASK, f, ns...);
+  }
```

Comment:
Copying will occur here; I wonder if you want to take `Ns` by universal reference `(Ns&& ... ns)` and then std::forward.